tsl/src/continuous_aggs/common.c

```diff
@@ -688,146 +683,73 @@ cagg_validate_query(const Query *query, const bool finalized, const char *cagg_s
 	}
 
-	/* Check if there are only two tables in the from list. */
-	fromList = query->jointree->fromlist;
-	if (list_length(fromList) > CONTINUOUS_AGG_MAX_JOIN_RELATIONS)
-	{
-		ereport(ERROR,
-				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-				 errmsg("only two tables with one hypertable and one normal table "
-						"are allowed in continuous aggregate view")));
-	}
-	/* Extra checks for joins in Caggs. */
-	if (list_length(fromList) == CONTINUOUS_AGG_MAX_JOIN_RELATIONS ||
-		!IsA(linitial(query->jointree->fromlist), RangeTblRef))
+	int num_tables = 0, num_hypertables = 0;
+	ListCell *lc;
+	foreach (lc, query->rtable)
 	{
-		if (list_length(fromList) == CONTINUOUS_AGG_MAX_JOIN_RELATIONS)
-		{
-			if (!IsA(linitial(fromList), RangeTblRef) || !IsA(lsecond(fromList), RangeTblRef))
-				ereport(ERROR,
-						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						 errmsg("invalid continuous aggregate view"),
-						 errdetail(
-							 "From clause can only have one hypertable and one normal table.")));
+		RangeTblEntry *inner_rte = lfirst_node(RangeTblEntry, lc);
 
-			rtref = linitial_node(RangeTblRef, query->jointree->fromlist);
-			rte = list_nth(query->rtable, rtref->rtindex - 1);
-			rtref_other = lsecond_node(RangeTblRef, query->jointree->fromlist);
-			rte_other = list_nth(query->rtable, rtref_other->rtindex - 1);
-			jointype = rte->jointype || rte_other->jointype;
-
-			if (query->jointree->quals != NULL && IsA(query->jointree->quals, OpExpr))
-				op = (OpExpr *) query->jointree->quals;
-		}
-		else
+		if (inner_rte->rtekind == RTE_RELATION)
 		{
-			ListCell *l;
-			foreach (l, query->jointree->fromlist)
+			if (ts_is_hypertable(inner_rte->relid) ||
+				ts_continuous_agg_find_by_relid(inner_rte->relid))
 			{
-				Node *jtnode = (Node *) lfirst(l);
-				JoinExpr *join = NULL;
-				if (IsA(jtnode, JoinExpr))
-				{
-					join = castNode(JoinExpr, jtnode);
-					jointype = join->jointype;
-					op = (OpExpr *) join->quals;
-					rte = list_nth(query->rtable, ((RangeTblRef *) join->larg)->rtindex - 1);
-					rte_other = list_nth(query->rtable, ((RangeTblRef *) join->rarg)->rtindex - 1);
-					if (rte->subquery != NULL || rte_other->subquery != NULL)
-						ereport(ERROR,
-								(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-								 errmsg("invalid continuous aggregate view"),
-								 errdetail("Sub-queries are not supported in FROM clause.")));
-					RangeTblEntry *jrte = rt_fetch(join->rtindex, query->rtable);
-					if (jrte->joinaliasvars == NIL)
-						ereport(ERROR,
-								(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-								 errmsg("invalid continuous aggregate view")));
-				}
+				num_hypertables++;
+				if (rte == NULL)
+					rte = copyObject(inner_rte);
 			}
+			else
+			{
+				num_tables++;
+				if (inner_rte->relkind == RELKIND_VIEW)
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("invalid continuous aggregate view"),
+							 errdetail("Views are not supported in continuous aggregates.")));
```

Comment:
Why are we blocking views?

Reply:
There's no techical problem to allow it but we always blocked it before and for now we'll continue blocking it as well. We can do a follow PR in the future expanding it a bit.

Reply:
I think we can just allow. These kind of restrictions only serve to prevent users from using the feature and are easily circumvented anyway. Less restrictions equals better :). On the technical side nothing changes by allowing this and with FDW tables it's easy to circumvent.

Reply:
Make sense. Removed the restriction